api/src/main/java/ai/djl/training/listener/EarlyStoppingListener.java

```diff
@@ -0,0 +1,175 @@
+ */
+public class EarlyStoppingListener implements TrainingListener {
+    private final double objectiveSuccess;
+
+    /**
+     * after minimum # epochs, consider stopping if:
+     */
+    private int minEpochs;
+    /**
+     * too much time elapsed
+     */
+    private int maxMinutes;
+    /**
+     * consider early stopping if not x% improvement
+     */
+    private double earlyStopPctImprovement;
+    /**
+     * stop if insufficient improvement for x epochs in a row
+     */
+    private int epochPatience;
+
+    private long startTimeMills;
+    private double prevLoss;
+    private int numberOfEpochsWithoutImprovements;
+
+    public EarlyStoppingListener() {
+        this.objectiveSuccess = 0;
+        this.minEpochs = 0;
+        this.maxMinutes = Integer.MAX_VALUE;
+        this.earlyStopPctImprovement = 0;
+        this.epochPatience = 0;
+    }
+
+    public EarlyStoppingListener(double objectiveSuccess, int minEpochs, int maxMinutes, double earlyStopPctImprovement, int earlyStopPatience) {
+        this.objectiveSuccess = objectiveSuccess;
+        this.minEpochs = minEpochs;
+        this.maxMinutes = maxMinutes;
+        this.earlyStopPctImprovement = earlyStopPctImprovement;
+        this.epochPatience = earlyStopPatience;
+    }
+
+    @Override
+    public void onEpoch(Trainer trainer) {
+        int currentEpoch = trainer.getTrainingResult().getEpoch();
+        // stopping criteria
+        final double loss = getLoss(trainer);
+        if (loss < objectiveSuccess) {
+            throw new EarlyStoppedException(currentEpoch, String.format("validation loss %s < objectiveSuccess %s", loss, objectiveSuccess));
+        }
+        if (currentEpoch >= minEpochs) {
+            double elapsedMinutes = (System.currentTimeMillis() - startTimeMills) / 60_000.0;
+            if (elapsedMinutes >= maxMinutes) {
+                throw new EarlyStoppedException(currentEpoch, String.format("Early stopping training: %.1f minutes elapsed >= %s maxMinutes", elapsedMinutes, maxMinutes));
+            }
+            // consider early stopping?
+            if (Double.isFinite(prevLoss)) {
+                double goalImprovement = prevLoss * (100 - earlyStopPctImprovement) / 100.0;
+                boolean improved = loss <= goalImprovement; // false if any NANs
+                if (improved) {
+                    numberOfEpochsWithoutImprovements = 0;
+                } else {
+                    numberOfEpochsWithoutImprovements++;
+                    if (numberOfEpochsWithoutImprovements >= epochPatience) {
+                        throw new EarlyStoppedException(currentEpoch, String.format("failed to achieve %s%% improvement %s times in a row",
+                                earlyStopPctImprovement, epochPatience));
+                    }
+                }
+            }
+        }
+        if (Double.isFinite(loss)) {
+            prevLoss = loss;
+        }
+    }
+
+    private static double getLoss(Trainer trainer) {
```

Comment:
Maybe have `TrainingResult result - trainer.getTrainingResult()` in `onEpoch`. Then, you can make this `getLoss(TrainingResult result)`

Reply:
`getLoss` now taks the TrainingResult as a param